tsdb/series_id.go

```diff
@@ -0,0 +1,74 @@
+package tsdb
+
+import (
+	"unsafe"
+
+	"github.com/influxdata/influxdb/models"
+)
+
+const (
+	// constants describing bit layout of id and type info
+	seriesIDTypeFlag  = 1 << 63                   // a flag marking that the id contains type info
+	seriesIDValueMask = 0xFFFFFFFF                // series ids numerically are 32 bits
+	seriesIDTypeShift = 32                        // we put the type right after the value info
+	seriesIDTypeMask  = 0xFF << seriesIDTypeShift // a mask for the type byte
+)
+
+// SeriesIDHasType returns if the raw id contains type information.
+func SeriesIDHasType(id uint64) bool { return id&seriesIDTypeFlag > 0 }
+
+// SeriesID is the type of a series id.
+type SeriesID struct{ ID uint64 }
```

Comment:
What's the benefit of using a struct instead of:

```go
type SeriesID uint64
```

Reply:
It doesn't automatically convert from an untyped constant. The only reason the ID field is actually exported is because a ton of tests were using some reflect business that didn't handle the unexported field. I wanted to be totally opaque so that I caught all usages of it. Many places were directly comparing to 0, or using <.

Reply:
@zeebo seems like a sensible choice. I think it might be worth just documenting that on the type so it doesn't get changed back by someone with good intentions.